### ISSUE TITLE
knife poisons stdout, making pipelining hard

### ISSUE THREAD

**Issue by eropple** [chef/chef#2980]
`knife` is fond of emitting non-actionable data to stdout in some cases. To wit:

```bash
$ knife search node -i -F text 'tags:zookeeper'
3 items found

zookeeper.test-cloud.infra.edddbb17
zookeeper.test-cloud.infra.ff2adb0f
zookeeper.test-cloud.infra.9095d1bf
```

This forces a consuming user to blunder around with `head` or `grep` to get something I can pipeline to other commands. Idiomatic Unix commands generally emit something like that header row (if they emit a header row at all) to stderr. If a user wishes to combine the two for output, `2>&1` is right there for her use--the opposite is not true.

`knife node list` doesn't exhibit the same behavior, which is good.

**Comment by danielsdeleo** [comment]
A bit of backstory. In the olden times, `knife` just reprinted the JSON data from the server for all of its commands. This was clearly not very readable. I spent a lot of time trying to find a text based format that would be nice and unix-y for pipelining to commands like `awk` and `cut`, but quite often you're dealing with deeply nested JSON data structures and there is simply no way to make them friendly for pipelining and human readable at the same time. So I opted to optimize the output for human readability and provide a handful of alternative data formats that could be used for interchanging data to other programs. Take, for example the output of `knife node show NODE -m`; it uses indentation to show context, which is easy for humans to look at, but actively hinders line-based text processing.

That said, if you want to move the "N items found\n\n" to stderr, that'd be fine, but at some point you're gonna hit a brick wall with the output format we have now. A 'line' formatter that works well with `cut` and `awk` and friends would be super awesome. I've always wanted something like that but haven't had the time to write it.

**Comment by eropple** [comment]
Hey Daniel--thanks for the reply, and I appreciate the historical lesson. I get the decision from your perspective, but it makes life a lot harder for actually using Chef for nontrivial tasks. Having to context-switch into Python or Ruby for things that are otherwise pretty easy to do in a straightforward shell script is a real drag on productivity and automation for Chef.

I'd love to contribute, but I'm communicating "this would make Chef more suitable for us long-term" for an organization who's paying Chef for Hosted Chef; my ability to contribute back to Chef for things that are not immediate and hard blockers is constrained and I have projects of my own for my free time. :-(

**Comment by danielsdeleo** [comment]
Ok, I patched knife so the count goes to stderr, so you can pipeline stdout more easily. I think I missed the cutoff for Chef 12.1, so this might not be released until 12.2.

We'd still be really interested in adding a formatter that provides line-based output for everything (like the `knife node show NODE -m` example) but realistically this would probably need to be implemented by an outside contributor given our current backlog and priorities.
